src/daniil_efremov_ex/ex06/logistic_regression.py

```diff
@@ -0,0 +1,307 @@
+# -*- coding: utf-8 -*-
+
+__author__ = 'Daniil Efremov'
+__email__ = 'daniil.vitalevich.efremov@nmbu.no'
+
+
+import numpy as np
+from sklearn.base import BaseEstimator, ClassifierMixin
+from sklearn.exceptions import NotFittedError
+from sklearn.utils import check_random_state, check_X_y
+
+
+def sigmoid(z):
+    """Perform a logistic transform on the input.
+    This function applies the sigmoidal function element-wise to all
+    elements of `z`. The sigmoidal function is on the following form:
+    .. math::
+        \frac{1}{1 + exp(-\mathbf{z})}.
+    Parameters
+    ----------
+    z : np.ndarray
+        Logit to transform.
+    Returns
+    -------
+    sigmoidal_transformed_z : np.ndarray
+        Transformed input.
+    """
+    sigmoidal_transformed_z = 1 / (1 + np.exp(-z))
+
+    return sigmoidal_transformed_z
+
+
+def predict_proba(coef, X):
+    r"""Predict the class probabilities for each data point in :math:`X`.
+    Estimate which class each data point in X corresponds to. This is done
+    according to the following formula.
+    .. math::
+        \hat{y}_i = \sigma(\mathbf{x}_i^T \mathbf{w}),
+    where :math:`x_i` is the i-th row in :math:`X` and :math:`\sigma` is
+    the sigmoidal function. Alternatively, in matrix-vector form:
+    .. math::
+        \hat{\mathbf{y}} = \sigma(X \mathbf{w}).
+    Parameters
+    ----------
+    coef : np.ndarray(shape=(r,))
+        The weight vector, :math:`w`
+    X : np.ndarray(shape=(n, r))
+        The data matrix (aka design or measurement matrix)
+    Returns
+    -------
+    p : np.ndarray(shape(n,))
+        The predicted class probabilities.
+    """
+    p = sigmoid(np.dot(X, coef))
+
+    return p
+
+
+def logistic_gradient(coef, X, y):
+    r"""Returns the gradient of a logistic regression model.
+    The gradient is given by
+    .. math::
+        \nabla_w L(\mathbf{w}; X,
+        \mathbf{y}) = \sum_i \mathbf{x}_i (y_i - \hat{y}_i), or, elementwise,
+    .. math::
+        \left[\nabla_w L(\mathbf{w}; X, \mathbf{y})\right]_j =
+        \frac{\partial L}{\partial w_j}
+        = \sum_i X_{ij} (y_i - \hat{y}_i),
+    where :math:`\hat{y}_i` is the predicted value for data point
+    :math:`i` and is given by :math:`\sigma(x_i^Tw)`, where
+    :math:`\sigma(z)` is the sigmoidal function.
+    Parameters
+    ----------
+    coef : np.ndarray(shape=(r,))
+        The weight vector, :math:`w`
+    X : np.ndarray(shape=(n, r))
+        The data matrix (aka design or measurement matrix)
+    y : np.ndarray(shape=(n,))
+        The true class labels for each data point.
+    Returns
+    -------
+    gradient : np.ndarray(shape=(r,))
+        The gradient of the cross entropy loss related to the linear
+        logistic regression model.
+    """
+
+    y_hat = predict_proba(coef, X)
+    something = y - y_hat
```

Comment:
This line makes the main code not able to reach the best accuracy when training.
It supposed to be `y_hat - y`. Also, don't use vague name like `something`, from the formula, you can name it `errors`, `residual_error`, `delta_y`.